src/Simplex/Messaging/Server/MsgStore/Journal.hs

```diff
@@ -0,0 +1,556 @@
+        (ws, wh) <- case writeHandle hs of
+          Nothing | msgCount writeState >= maxMsgCount -> switchWriteJournal hs
+          wh_ -> pure (writeState, fromMaybe (readHandle hs) wh_)
+        let msgCount' = msgCount ws + 1
+            ws' = ws {msgPos = msgPos ws + 1, msgCount = msgCount', bytePos = bytePos ws + msgLen}
+            rs' = if journalId ws == journalId rs then rs {msgCount = msgCount'} else rs
+            !st' = st {writeState = ws', readState = rs', canWrite = canWrt', size = size + 1}
+        when (size == 0) $ atomically $ writeTVar (tipMsg q) $ Just (Just (msg, msgLen))
+        hAppend wh msgStr
+        updateQueueState q logState hs st'
+        where
+          createQueueDir = do
+            createDirectoryIfMissing True queueDirectory
+            let statePath = queueDirectory </> (queueLogFileName <> logFileExt)
+            sh <- openFile statePath AppendMode
+            B.hPutStr sh ""
+            rh <- createNewJournal queueDirectory $ journalId rs
+            let hs = MsgQueueHandles {stateHandle = sh, readHandle = rh, writeHandle = Nothing}
+            atomically $ writeTVar handles $ Just hs
+            pure hs
+          switchWriteJournal hs = do
+            journalId <- newJournalId random
+            wh <- createNewJournal queueDirectory journalId
+            atomically $ writeTVar handles $ Just $ hs {writeHandle = Just wh}
+            pure (newJournalState journalId, wh)
+
+  getQueueSize :: JournalMsgQueue -> IO Int
+  getQueueSize JournalMsgQueue {state} = size <$> readTVarIO state
+
+  tryPeekMsg_ :: JournalMsgQueue -> NonAtomicIO (Maybe Message)
+  tryPeekMsg_ q@JournalMsgQueue {tipMsg, handles} =
+    NonAtomicIO $ readTVarIO handles $>>= chooseReadJournal q True $>>= peekMsg
+    where
+      peekMsg (rs, h) = readTVarIO tipMsg >>= maybe readMsg (pure . fmap fst)
+        where
+          readMsg = do
+            -- TODO handle errors
+            s <- hGetLineAt h $ bytePos rs
+            -- TODO handle errors
+            Right msg <- pure $ strDecode s
+            atomically $ writeTVar tipMsg $ Just (Just (msg, B.length s + 1)) -- 1 is to account for new line
+            pure $ Just msg
+
+  tryDeleteMsg_ :: JournalMsgQueue -> NonAtomicIO ()
+  tryDeleteMsg_ q@JournalMsgQueue {tipMsg, handles} = NonAtomicIO $
+    void $
+      readTVarIO tipMsg -- if there is no cached tipMsg, do nothing
+        $>>= (pure . fmap snd)
+        $>>= \len -> readTVarIO handles
+        $>>= \hs -> updateReadPos q True len hs $> Just ()
+
+  atomicQueue :: JournalMsgQueue -> String -> NonAtomicIO a -> ExceptT ErrorType IO a
+  atomicQueue mq op (NonAtomicIO a) = tryStore op $ withLock' (queueLock mq) op $ a
+
+tryStore :: String -> IO a -> ExceptT ErrorType IO a
+tryStore op a =
+  ExceptT $
+    (Right <$> a) `catchAny` \e ->
+      let e' = op <> " " <> show e
+       in logError ("STORE ERROR " <> T.pack e') $> Left (STORE e')
```

Comment:
this doesn't seem to guarantee atomicity by itself.. if error happened after (mid) changes, they don't seem to be rolled back here..

Reply:
if it's re-using acid terminology `atomicQueue` really means `isolatedQueue` here I think

Reply:
done